### PR TITLE
DUPLO-11338 support ImageConfig for lambda functions

### PR DESCRIPTION
**Pull request by ddiaz-duplo**
## Change description

DUPLO-11338: Add support for `imageConfig` 
DUPLO-11626: Update v1 -> v3 apis for updating lambda function code and configuration

## Testing Evidence
Ran `terraform apply` on this resource:
```
resource "duplocloud_aws_lambda_function" "myfunction" {

  tenant_id   = "c7163b39-43ca-4d44-81ce-9a323087039b"
  name        = "myfunction"
  description = "A description of my function"

  package_type = "Image"
  image_uri = "151326636836.dkr.ecr.us-west-2.amazonaws.com/whatever:latest"

  environment {
    variables = {
      "FOO" = "BAR"
    }
  }

  image_config {
    command = ["echo", "hello world"]
    entry_point = ["echo hello workd"]
    working_directory = "/tmp"
  }

  timeout     = 60
  memory_size = 512
}
```

Ran `terraform state show duplocloud_aws_lambda_function.myfunction` to view state
```
╰─ terraform state show duplocloud_aws_lambda_function.myfunction   
# duplocloud_aws_lambda_function.myfunction:
resource "duplocloud_aws_lambda_function" "myfunction" {
    arn               = "arn:aws:lambda:us-west-2:151326636836:function:duploservices-default-myfunction"
    description       = "A description of my function"
    ephemeral_storage = 512
    fullname          = "duploservices-default-myfunction"
    id                = "c7163b39-43ca-4d44-81ce-9a323087039b/myfunction"
    image_uri         = "151326636836.dkr.ecr.us-west-2.amazonaws.com/whatever:latest"
    last_modified     = "2023-09-22T19:58:24.482+00:00"
    layers            = []
    memory_size       = 512
    name              = "myfunction"
    package_type      = "Image"
    role              = "arn:aws:iam::151326636836:role/duploservices-default"
    source_code_hash  = "1754443cbee58558c930df055ace7f1fd1224a6a4702277d21cc150a0f3a0b47"
    source_code_size  = 0
    tags              = {
        "TENANT_NAME"   = "default"
        "duplo-project" = "default"
    }
    tenant_id         = "c7163b39-43ca-4d44-81ce-9a323087039b"
    timeout           = 60
    version           = "$LATEST"

    environment {
        variables = {
            "FOO" = "BAR"
        }
    }

    image_config {
        command           = [
            "echo",
            "hello world",
        ]
        entry_point       = [
            "echo hello workd",
        ]
        working_directory = "/tmp"
    }

    tracing_config {
        mode = "PassThrough"
    }
}
```

run `terraform apply` again to make sure it recognizes nothing changed

```
╰─ terraform apply                                               
duplocloud_aws_lambda_function.myfunction: Refreshing state... [id=c7163b39-43ca-4d44-81ce-9a323087039b/myfunction]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

```

change property `working_directory` to /tmp2 and run `terraform apply` again

```
╰─ terraform apply
duplocloud_aws_lambda_function.myfunction: Refreshing state... [id=c7163b39-43ca-4d44-81ce-9a323087039b/myfunction]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

## Result

image

## Type of change
- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

### Development

- [x] Pull requests may not be submitted to the *master* branch (use *develop* instead) - or they will be closed.
- [x] Lint rules pass locally
- [x] Application changes have been tested thoroughly
- [x] Automated tests covering modified code pass

### Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### Code review 

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [x] Pull request linked to task tracker where applicable

### CODE REVIEW

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-11338 TF: support for Image configurations in Lambda](https://app.clickup.com/t/8655600/DUPLO-11338)

**Review by joek-duplo — COMMENTED** [review]
this looks good so far to me. :+1:

**Review by joek-duplo — CHANGES_REQUESTED** [review]

File: duplocloud/resource_duplo_aws_lambda_function.go [hunk 1]
```diff
@@ -497,6 +552,7 @@ func updateAwsLambdaFunctionConfig(tenantID, name string, d *schema.ResourceData
 		Timeout:      d.Get("timeout").(int),
 		MemorySize:   d.Get("memory_size").(int),
 		Tags:         expandAwsLambdaTags(d),
+		ImageConfig:  d.Get("image_config").(*duplosdk.DuploLambdaImageConfig),
```
[the remark this answers is not in the document]
Comment (ddiaz-duplo) [reply duplocloud/resource_duplo_aws_lambda_function.go:555]: you're right, I'll fix this. I added my testing above. I thought I had tested the update case but I might've changed the type in one of my commits.

[the remark this answers is not in the document]
Comment (ddiaz-duplo) [reply duplocloud/resource_duplo_aws_lambda_function.go]: Fixed.

**Review by joek-duplo — CHANGES_REQUESTED** [review]
nice work adding the test suite, btw!

question on the image config stuff:
- does the backend not support updates to image config?  if not, this PR should be good already.
- if the backend *DOES* support updates to image config, we need to support it here too

**Comment by ddiaz-duplo** [comment]
> nice work adding the test suite, btw!
> 
> question on the image config stuff:
> 
> * does the backend not support updates to image config?  if not, this PR should be good already.
> * if the backend _DOES_ support updates to image config, we need to support it here too

I’m only seeing one endpoint `subscriptions/{tenant_id}/UpdateLambdaFunctionConfiguration`
Action: `HttpPost` and thats the one used in the TF code.
[TF Code](https://github.com/duplocloud/terraform-provider-duplocloud/blob/515f2a43a7cec2a6e2442a94698f2aca07a69b9d/duplosdk/aws_lambda_function.go#L169)
[C# Code](https://github.com/duplocloud-internal/duplo/blob/00e8ae380d6d6c52ca7759ec18e4baf5f4a9c81a/ContainerManagement/Containers/Api/Controllers/ContainersTenantApiController.cs#L2163)

**Review by joek-duplo — APPROVED** [review]

**Pull request merged**